### PR TITLE
Type-only auto-import improvements

### LINKED ISSUE #53008: Auto-import will always use the first import instead of looking for the first import of the correct kind
# Bug Report

### 🔎 Search Terms

auto import type-only value

### 🕗 Version & Regression Information

- This is the behaviour in every version I tried

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBDAnmApnA3nAwhcEB2K+MAKsmgL5wBmUucA5FCgIYDGMDA3AFCiSwMcAK4BnFAGUYLGJRp0QjZu068eKAB4D41Yfg7AC2XJELEAFJjB0wcKgC4h1iGEc48Z0uTsBKDDzg4NgJReGCAExQASVEAVXx+aGliOABeEXEpGRRzAEYfNUCAeiKEKGAAcwqUKDgWYRgIAFpEwWpoBAALYFE4ABtgQgCMlABRamoUDnNzP1SAPgwKABo4AG0AXQKeCh4gA)

### 💻 Code

```ts
import type { ComponentType } from 'react';
import { useState } from 'react';

export function Component({ prop } : { prop: ComponentType }) {
  const codeIsUnimportant = useState(1);

  // trigger auto-import for this line
  useEffect(() => {}, []);
}
```

### 🙂 Expected behaviour

When I trigger auto-import for the `useEffect` function, TS should look at all of the imports and find the first **_value_** import declaration with the correct source to use.
I.e. I would expect this result:
```ts
import type { ComponentType } from 'react';
import { useState, useEffect } from 'react';
```

### 🙁 Actual behaviour

When I trigger auto-import for the `useEffect` function, TS looks for the first import declaration with the correct source, even if it's a `type`-only import. If it is `type`-only, TS will remove the qualifier entirely.

I.e. it results in this code
```ts
import { ComponentType, useEffect } from 'react';
import { useState } from 'react';
```

Which is pretty bad because it means TS changes the meaning of the code for some compilers!

If 5.0's `verbatimModuleSyntax` is turned on, then instead you get this:
```ts
import { useEffect, type ComponentType } from 'react';
import { useState } from 'react';
```
Which is *marginally* better as it at least preserves the meaning of the code.

---

Note that the same thing happens if you reverse the situation.
[playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgVwM4FMDKMCGN1wC+cAZlBCHAORTo4DGMVA3AFCiSxwwCeYBSAMIVIAO3SiYAFT4FiZCtVoMmbVugAeneCWSjGwCKLjDwRiTAAUSMOTBE4ALkRxbEMM9NiLM-kQCUiKxwcPRGqPBhACboAJKoAKqiHNC4knAAvCgY2HjolgCM-mohAPSl3FDAAObV6FBwOMgwEAC0KVwk0NwAFsCocAA2wOLB3LJwUplwsaIAbjjDUQCicxZshKxAA)

```ts
import { useState } from 'react';
import type { ComponentType } from 'react';

export function Component({ prop } : { prop: ComponentType }) {
  const codeIsUnimportant = useState(1);

  // trigger auto-import for this line
  type T = InvalidEvent;
}
```

By default TS will insert the type import in the value import:
```ts
import { InvalidEvent, useState } from 'react';
import type { ComponentType } from 'react';
```

If 5.0's `verbatimModuleSyntax` is turned on, then instead you get this:
```ts
import { type InvalidEvent, useState } from 'react';
import type { ComponentType } from 'react';
```

---

### Suggested change:

- When auto-importing, gather all of the import declaratinos.
- If the imported thing is a type, then prefer using the first type-only import declaration.
- If the imported thing is a value, then prefer using the first value import declaration.
- If there is no matching declaration, then I think one of the following should occur:
    - If the thing is a value and the only import declarations are type-only, then insert a new value import declaration. **DO NOT** mutate the existing import declaration(s) to remove the top-level type-only qualifier.
    - If the thing is a type and the only import declarations are value, then one of the following should occur (perhaps configurably?):
        - Insert a new specifier into the value declaration, qualifying the specifier with `type`
        - Insert a new type-only import declaration.

The most important bit, IMO, is that TS should ***NEVER*** remove the top-level `type` qualifier from an import when performing an auto-import.

### PR DESCRIPTION
**Pull request by andrewbranch**
Closes #53008

1. When more than one import declaration from the same module specifier exists, find the one that’s the best match with respect to type-onlyness instead of modifying the first one to work:
   ```ts
   import type {} from "react"; // <-- auto-imported types will go here
   import {} from "react";      // <-- auto-imported values will go here
   ```
2. When promoting a type-only import declaration to a real one due to a new value import, always add `type` modifiers to the existing specifiers, even if emit settings don’t require it. For example, if a value import of `SomeValue` is to be auto-imported here:
   ```ts
   import type { SomeType } from "mod";
   ```
   the result will now always be:
   ```ts
   import { SomeValue, type SomeType } from "mod";
   ```
   regardless of compiler options. Previously, we would only insert the `type` modifier on `SomeType` if it was necessary according to compiler options, e.g. if `verbatimModuleSyntax` is enabled.

### CODE REVIEW

File: tests/cases/fourslash/autoImportTypeOnlyPreferred2.ts [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+export function Component({ prop } : { prop: ComponentType }) {
+    const codeIsUnimportant = useState(1);
+    useEffect(() => {}, []);
+}`,
+]);
+
+goTo.marker("2");
+verify.importFixAtPosition([
+`import { useState } from "react";
+import type { ComponentProps, ComponentType } from "react";
```
Comment (jakebailey) [inline tests/cases/fourslash/autoImportTypeOnlyPreferred2.ts:38]: Why is the order here changing? I know we organize types below non-types (until I get around to improving that), but why is it happening in a fix when there are already imports present?

Comment (andrewbranch) [reply tests/cases/fourslash/autoImportTypeOnlyPreferred2.ts:38]: It didn’t, there are two files in this test with opposite orders. See the `goTo.marker("2")`

File: src/services/codefixes/importFixes.ts [hunk 2]
```diff
@@ -1364,13 +1386,11 @@ function doAddExistingFix(
...
-    // some of them don't strictly need to be marked type-only (if they have a value meaning and are
-    // never used in an emitting position). These are allowed to be imported without being type-only,
-    // but the user has clearly already signified that they don't need them to be present at runtime
-    // by placing them in a type-only import. So, just mark each specifier as type-only.
-    const convertExistingToTypeOnly = promoteFromTypeOnly && importNameElisionDisabled(compilerOptions);
+    // We used to convert existing specifiers to type-only only if compiler options indicated that
+    // would be meaningful (see the `importNameElisionDisabled` utility function), but user
+    // feedback indicated a preference for preserving the type-onlyness of existing specifiers
+    // regardless of whether it would make a difference in emit.
+    const convertExistingSpecifiersToTypeOnly = promoteFromTypeOnly;
```
Comment (jakebailey) [inline src/services/codefixes/importFixes.ts:1393]: Does this constant need to stay? It's already changing names, so it seems like the uses could just be updated to be `promoteFromTypeOnly`, unless there's something I'm missing that's not shown in the diff.

**Review by jakebailey — APPROVED** [review]

File: src/services/codefixes/importFixes.ts (hunk 2 above)
Comment (andrewbranch) [reply src/services/codefixes/importFixes.ts:1393]: It doesn’t, but looking at where it was used, I thought it was more clear to keep it.

**Pull request merged**